U4-11502 - fall-back languages [v8]

AndyButland opened the pull request:
In this PR I've set up the database, services and back-office editing screen to allow for the selection of a fall back language for each language, and then implemented and registered an instance of `IPublishedValueFallback` that uses this to provide fall back content if the content for the requested language is not found.

Back-office changes look like this:
![image](https://user-images.githubusercontent.com/1993459/42420746-936acfb4-82ca-11e8-9448-cac93b35a3c3.png)
![image](https://user-images.githubusercontent.com/1993459/42420747-9802cd56-82ca-11e8-9cfa-0103a7bc0a83.png)

Database changes have a migration which should replicate the following SQL:

```
ALTER TABLE umbracoLanguage
ADD fallbackLanguageId int NULL
GO

ALTER TABLE umbracoLanguage WITH CHECK ADD  CONSTRAINT FK_umbracoLanguage_umbracoLanguage_fallbackLanguageId FOREIGN KEY(fallbackLanguageId)
REFERENCES umbracoLanguage (id)
GO

CREATE INDEX IX_umbracoLanguage_fallbackLanguageId
ON umbracoLanguage (fallbackLanguageId)
GO
```

With this in place, you should be able to configure that the Italian language falls back to English, and then if you request `@Model.Value("welcomeText", "it")` - and no Italian content is published (or the value for the requested property is empty), you'll get back the English content (if that's published of course).

It's set up to allow for chains of fall back language - so Spanish could fall back to Italian and then to English for example.

And also should work in conjunction with the existing fall back behaviour that recurses up the tree; I've introduced a new enum and parameter to allow the caller to indicate whether the preference for the first attempt to fall back is via language or tree recursion.

I'm concious of a lack of unit tests around the key class of `PublishedValueLanguageFallback`, but ran into some trouble trying to set these up.  Ideally would do so as some additional tests within `PublishedContentTests`, but that doesn't seem to have access to services that can be used to save values, and is currently based off (I assume obsolete?) XML content.  So that might need some work still or some pointers that I could pick up. 

So currently all testing has been manual.

Something else that might need some thought is whether there should be a way for the developer to request to use this behaviour or not, per property value request in a template - in the same way you currently pass the `recurse` parameter if you want to fall back to content up the tree.  Currently I've not done this, and considered if a developer sets up fall-back languages in the back-office, they'll be used for all property value requests.

zpqrtbnk commented:
Hey - impressive work. Going to review in more details hopefully within the coming 2 weeks.

Few things: migrations now need to be registed in `UmbracoPlan` - ping me if that does not make any sense. Also, if you look into `PublishedContentMoreTests` you'll see that there is a `CreatePublishedSnapshot` which kinda creates a complete content cache (without resorting to services). Properties there do not support variations yet but could be extended I guess.

Shazwazza commented on src/Umbraco.Core/Models/ILanguage.cs:
```diff
@@ -33,5 +33,11 @@ public interface ILanguage : IEntity, IRememberBeingDirty
+        /// Defines the fallback language that can be used in multi-lingual scenarios to provide
+        /// content if the requested language does not have it published.
+        /// </summary>
+        ILanguage FallbackLanguage { get; set; }
```
We should make this an `int?` and not have a direct reference to the other ILanguage. This will also simplify a few things with this implementation. The primary reason is because we are slowly trying to reduce the object graphs that get cached. All languages get cached and this will mean there's actually more memory consumed by the entire cache by having another object reference. (hope that makes sense?)

AndyButland replied:
Yes, makes sense.  As you say, it does simplify things too.  Will take a look at updating to this.

Shazwazza commented on src/Umbraco.Core/Persistence/Dtos/LanguageDto.cs:
```diff
@@ -38,5 +38,15 @@ internal class LanguageDto
+        /// </summary>
+        [Column("fallbackLanguageId")]
+        [ForeignKey(typeof(LanguageDto), Column = "id")]
+        [Index(IndexTypes.NonClustered)]
```
No need to have an index on this column since we never query by this column

AndyButland replied:
Removed.  Initially thought we might use if for a self-join, but wasn't needed.

AndyButland commented:
Updates made now @Shazwazza

Shazwazza commented:
@AndyButland nice! and looks a bit simpler too with less code 🎉 @zpqrtbnk do you have any other recommendations?

zpqrtbnk commented:
Dealing with conflicts, and reviewing now.

zpqrtbnk commented:
So... I have a commit dealing with conflicts and with a few fixes - if you allow me I can push to your branch.

Now, after reviewing with Shannon, some comments:
* Deleting a language which is a fallback for another language, is prevented by the controller, but doing it directly through the service is throwing an exception. Can we update the language repository so that it first unregisters the deleted language from all fallbacks, so then it can be deleted?
* The fallback logic does not detect loops so it could throw the site into an infinite loop. Can we keep track of visited languages and throw (or just abort?) in case of a loop?
* Should this validation also happen whenever a language is saved? And maybe raise an error in the UI? That might be complex to implement and require some changes at service level. Maybe we don't need it.
* The fallback methods have both the `bool recurse` parameter and the new fallback priority - wondering whether they should be merged, with the priority having a default "nothing" value (so, "nothing", "recurse", "fallback").
* More generally, what about other priorities? Could ppl want to write their own fallback provider *and* pass a different priority? Instead of being limited to the enum values? This is an open discussion, we thought about passing an integer (and defining constants), or anything that can be extended?

Oh and I forgot: loving it ;-)

AndyButland commented:
Some updates and response to comments:

- Have allowed delete of a language referenced by others as a fall-back by setting references to null (i.e. removing the relationship).  Because of this I've put back the index on the field that I originally put in and removed following earlier comments, as it would be used here I think.  Obviously see what you think though, can remove again if you don't think it's adding enough value.
- I've set up a check for loops, that aborts if we attempt to check the a language a second time.
- As noted there's validation to prevent the creation of loops in the controller.  Perhaps that could be in the service but actually it's consistent here with other checks that are being made (such as trying to delete the default language), so think maybe OK as is.  Someone could create a looping fall-back language via a service, but as there's a check now for loops, this isn't so critical.

_The last two I'm planning to get to later, so still have a bit of work to do before reviewing again or accepting..._

AndyButland commented:
Based on the last couple of comments I've rejigged it to combine the parameters for recursive search and fallback priority into one, an array of integers each of which defines via a constant one or more "fall back methods", in order of priority.  I think that makes things consistent - in the sense that for my initial version (as for v7) you have to pass the `recurse` boolean parameter to request falling back up the tree, but for the language fall-back it just happened automatically.  This way, you specify what you want explicitly.

It's also flexible in that a developer can define other methods, via a new constant integer, that they can use in their own implementation of `IPublishedValueFallback`.  And as it's an array, they can order the methods to determine priority any way they want.

I did look to  created an single integer overload to simplify the API when you only want to use one method, but unfortunately that runs into ambiguous method compilation errors.

So we end up with:

- `@Model.Value("bodyText", "es")` - returns the Spanish value, with no fall-back
- `@Model.Value("bodyText", fallbackMethods: new[] { Constants.Content.FallbackMethods.RecursiveTree })` - uses recursive parent node fall back
- `@Model.Value("bodyText", "es", fallbackMethods: new[] { Constants.Content.FallbackMethods.FallbackLanguage })` - uses fall back languages
- `@Model.Value("bodyText", "es", fallbackMethods: new[] { Constants.Content.FallbackMethods.FallbackLanguage, Constants.Content.FallbackMethods.RecursiveTree })` - uses fall back languages and then if still not found uses recursive parent node
- `@Model.Value("bodyText", "es", fallbackMethods: new[] { Constants.Content.FallbackMethods.RecursiveTree, Constants.Content.FallbackMethods.FallbackLanguage })` - uses recursive parent node and then if still not found uses fall back languages

zpqrtbnk commented:
Going to look at this ;-)

zpqrtbnk commented:
As discussed during our chat, have pushed the `temp8` merge, and some refactoring. But there are still some open questions, the whole discussion being detailed in comments in `IPublishedValueFallback`. The whole thing still needs to mature, so that using fallbacks, and default values, etc. is as easy & simple it possible for ppl creating front-end views - while extensible enough.

AndyButland commented:
Think something may have gone awry here @zpqrtbnk - with your update the PR is enormous now?

AndyButland commented:
Have had a look over the latest, and so on the comments in code:

- _"what's with visitedLanguage"_ - I felt I needed this originally due to the recursive nature of how I'd implemented the fallback, but prefer your implementation, and then as you say, it's not needed.
- _"so chaining things through an array of ints is not... convenient"_ - yes, agreed.  I was looking here to try to find a way to allow people to control the order by which fall back methods are tried (e.g. language first, then recurse up the tree; or vice versa).  But integer values allow this extension too.
- _"it feels like ppl could have ValueFallback.LanguageAndRecurse"_ - they could, or this could be in core.  It seemed to be fairly straightforward, so I've implemented it, which then restores the functionality that was available with the "array of integers" method discussed in the point above.
- _"currently, no policies on `IPublishedProperty` nor `IPublishedElement`, but some may apply (language)"_ - I've added these (though not got them under unit tests like for `IPublishedContent`).  Note that `Value<T>` on `PublishedPropertyExtension` wasn't previously calling `PublishedValueFallback`, so I've added this call.  It would be nice to DRY up the logic in the now 3 versions of `TryGetValueWithLanguageFallback` in `PublishedValueFallback`, but without a common interface - which can't create as some of the methods used are extension methods - couldn't find a nice way of doing this.

After that comments get a bit beyond where I can comfortably go I think, so will leave it there for now.  Look forward to seeing how it progresses.

zpqrtbnk commented:
so... I went through several different solutions and came back to passing an array of integers to indicate the fallback strategies - only, have added some syntactic sugar so you can write

    model.Value<string>("title", fallback: Fallback.ToLanguage)
    model.Value<string>("title", fallback: Fallback.To(Fallback.Language, Fallback.Recurse))
    model.Value<string>("title", "de-De", fallback: Fallback.ToDefaultValue, "someDefaultValue")

'course MB will support it

    model.Value(x => x.Title, fallback: Fallback.ToLanguage)

etc

but it's still syntactic sugar, providing an array of ints in a (as) simple (as possible) way.

been testing it and fixing things here and there... and now... guess it's time to merge!

WOOT!

AndyButland commented:
Nice, I like those syntax touches.  Thanks for persevering with it; glad it's made it in.
